macros/src/validation.rs

```diff
@@ -114,37 +120,72 @@ fn validate_guard_signatures(sm: &ParsedStateMachine) -> Result<(), parse::Error
-            {
-                let signature = FunctionSignature::new_guard(in_state_data, event_data, *is_async);
-
-                // If the action is not yet known, add it to our tracking list.
-                guards
-                    .entry(guard.to_string())
-                    .or_insert_with(|| signature.clone());
-
-                // Check that the call signature is equivalent to the recorded signature for this
-                // guard.
-                if guards.get(&guard.to_string()).unwrap() != &signature {
-                    return Err(parse::Error::new(
-                        Span::call_site(),
-                        format!("Guard `{}` can only be reused when all input states and events have the same data", guard),
-                    ));
+            for transition in &event_mapping.transitions {
+                if let Some(guard_expression) = &transition.guard {
+                    let res = visit_guards(guard_expression, |guard| {
+                        let signature =
+                            FunctionSignature::new_guard(in_state_data, event_data, guard.is_async);
+
+                        // If the action is not yet known, add it to our tracking list.
+                        guards
+                            .entry(guard.ident.to_string())
+                            .or_insert_with(|| signature.clone());
+
+                        // Check that the call signature is equivalent to the recorded signature for this
+                        // guard.
+                        if guards.get(&guard.ident.to_string()).unwrap() != &signature {
+                            return Err(parse::Error::new(
+                                Span::call_site(),
+                                format!("Guard `{}` can only be reused when all input states and events have the same data", guard.ident),
+                            ));
+                        }
+                        Ok(())
+                    });
+                    res?;
                 }
             }
         }
     }
 
     Ok(())
 }
+fn validate_unreachable_transitions(sm: &ParsedStateMachine) -> Result<(), parse::Error> {
+    let all_transitions = &sm.states_events_mapping;
+    for (in_state, event_mappings) in all_transitions {
+        for (event, event_mapping) in event_mappings {
+            // more than single transition for (in_state,event)
+            if event_mapping.transitions.len() > 1 {
+                let mut unguarded_count = 0;
+                for t in &event_mapping.transitions {
+                    if let Some(g) = &t.guard {
+                        if unguarded_count > 0 { // Guarded transition AFTER an unguarded one
+                            return Err(parse::Error::new(
+                                Span::call_site(),
+                                format!("{} + {}: [{}] : guarded transition is unreachable because it follows an unguarded transition, which handles all cases",
```

Comment:
This seems like a key piece of information that needs to be documented somewhere (i.e. that guard definition ordering matters and they are evaluated in definition order)